compression.go

```diff
@@ -0,0 +1,40 @@
+package kafka
+
+const (
+	CompressionNone int8 = iota
+	CompressionGZIP
+	CompressionSnappy
+	CompressionLZ4
+)
+
+type CompressionCodec struct {
+	str    func() string
+	encode func(src []byte, level int) ([]byte, error)
+	decode func(src []byte) ([]byte, error)
+}
```

Comment:
There's a weird dependency on the gzip compression level exposed in this abstraction. Neither lz4 nor snappy have such configuration.

How about something like this:
```go
package kafka

type CompressionCodec interface {
    Encode(dst, src []byte) (int, error)
    Decode(dst, src []byte) (int, error)
}

type Message struct {
    ...
    CompressionCodec CompressionCodec // no need to carry extra options if we carry the codec directly
}

func RegisterCompressionCodec(code int8, newCodec func() CompressionCodec) {
    ...
}

func NewCompressionCodec(code int8) CompressionCodec {
    // lookup the codec registered with `code`
    ...
}
```
```go
package gzip

type CompressionCodec struct {
    // use the zero-value, then initialize with the Reset method in the Encode/Decode
    r gzip.Reader
    w gzip.Writer
}

func NewCodec() *CompressionCodec {
    return NewCodec(gzip.DefaultCompression)
}

func NewCodecLevel(level int) *CompressionCodec {
    ...
}

const Code = 1 // gzip compression codec code in the kafka protocol

func init() {
    kafka.RegisterCompressionCodec(Code, func() kafka.CompressionCodec {
        return NewCodec()
    })
}
```
Here are a couple of reasons why I'm suggesting those changes:
- the compression level on gzip is only necessary on the encoder side, and the codec registry is mostly useful for decompression, so no reason to expose those implementation details in the generic interface
- using `Encode(dst, src []byte) (int, error)` instead of `Encode(src []byte) ([]byte, error)` means that the API gives an opportunity for optimizing the allocation of the destination buffer, which is otherwise not possible if the codec is in charge of allocating it.
- using the kafka codec ID instead of a codec name means we're not re-inventing yet another way to identify the compression codec, we're just using the values that are defined by the protocol.
- in each sub-package providing the implementation of a compression codec, the codec constructors return a concrete type instead of an interface, this is considered a best practice in Go (that's what the `compress/gzip` package does for example).